crates/bevy_ecs/src/schedule/graph/graph_map.rs

```diff
@@ -0,0 +1,480 @@
+
+use bevy_utils::{hashbrown::HashSet, AHasher};
+use core::{
+    fmt,
+    hash::{BuildHasher, BuildHasherDefault, Hash},
+};
+use indexmap::IndexMap;
+
+use super::{tarjan_scc::TarjanScc, NodeId};
+
+use Direction::{Incoming, Outgoing};
+
+/// A `Graph` with undirected edges.
+///
+/// For example, an edge between *1* and *2* is equivalent to an edge between
+/// *2* and *1*.
+pub type UnGraph<S = BuildHasherDefault<AHasher>> = Graph<false, S>;
+
+/// A `Graph` with directed edges.
+///
+/// For example, an edge from *1* to *2* is distinct from an edge from *2* to
+/// *1*.
+pub type DiGraph<S = BuildHasherDefault<AHasher>> = Graph<true, S>;
+
+/// `Graph<DIRECTED>` is a graph datastructure using an associative array
+/// of its node weights `NodeId`.
+///
+/// It uses an combined adjacency list and sparse adjacency matrix
+/// representation, using **O(|V| + |E|)** space, and allows testing for edge
+/// existence in constant time.
+///
+/// `Graph` is parameterized over:
+///
+/// - Constant generic bool `DIRECTED` determines whether the graph edges are directed or
+///   undirected.
+/// - The `BuildHasher` `S`.
+///
+/// You can use the type aliases `UnGraph` and `DiGraph` for convenience.
+///
+/// `Graph` does not allow parallel edges, but self loops are allowed.
+#[derive(Clone)]
+pub struct Graph<const DIRECTED: bool, S = BuildHasherDefault<AHasher>>
+where
+    S: BuildHasher,
+{
+    nodes: IndexMap<NodeId, Vec<CompactNodeIdAndDirection>, S>,
+    edges: HashSet<CompactNodeIdPair, S>,
+}
+
+impl<const DIRECTED: bool, S: BuildHasher> fmt::Debug for Graph<DIRECTED, S> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.nodes.fmt(f)
+    }
+}
+
+impl<const DIRECTED: bool, S> Graph<DIRECTED, S>
+where
+    S: BuildHasher,
+{
+    /// Create a new `Graph`
+    pub(crate) fn new() -> Self
+    where
+        S: Default,
+    {
+        Self::default()
+    }
+
+    /// Create a new `Graph` with estimated capacity.
+    pub(crate) fn with_capacity(nodes: usize, edges: usize) -> Self
+    where
+        S: Default,
+    {
+        Self {
+            nodes: IndexMap::with_capacity_and_hasher(nodes, S::default()),
+            edges: HashSet::with_capacity_and_hasher(edges, S::default()),
+        }
+    }
+
+    /// Use their natural order to map the node pair (a, b) to a canonical edge id.
+    #[inline]
+    fn edge_key(a: NodeId, b: NodeId) -> CompactNodeIdPair {
+        let (a, b) = if DIRECTED || a <= b { (a, b) } else { (b, a) };
+
+        CompactNodeIdPair::store(a, b)
+    }
+
+    /// Return the number of nodes in the graph.
+    pub fn node_count(&self) -> usize {
+        self.nodes.len()
+    }
+
+    /// Add node `n` to the graph.
+    pub(crate) fn add_node(&mut self, n: NodeId) {
+        self.nodes.entry(n).or_default();
+    }
+
+    /// Remove a node `n` from the graph.
+    ///
+    /// Computes in **O(V)** time, due to the removal of edges with other nodes.
```

Comment:
```suggestion
    /// Computes in **O(N)** time, due to the removal of edges with other nodes.
```
same here; V is for graph theory, but N is closer to the code.